### DISCUSSION TITLE
JSONB supported as of GRDB 7

### DISCUSSION THREAD

**Discussion by Jason-Abbott** [groue/GRDB.swift#1656] [Show and tell]
**Important caveat**: this requires the [GRDB custom build](https://github.com/groue/GRDB.swift/blob/master/Documentation/CustomSQLiteBuilds.md) process since it depends on SQLite 3.45.0 or higher.

The benefits of `BLOB` rather than `TEXT`-backed JSON, according to the SQLite [documentation](https://sqlite.org/jsonb.html), are "that JSONB is both slightly smaller (by between 5% and 10% in most cases) and can be processed in less than half the number of CPU cycles.”

I was already using a custom build for other reasons so when I saw @groue had updated [SQLiteLib](https://github.com/swiftlyfalling/SQLiteLib) as part of GRDB 7, it was easy to start using JSONB.

## First verifying with manual SQLite changes 

I first vetted the change with manual updates to a test SQLite database. The steps I took were
- Add a `<column>_new` variant of type `BLOB` corresponding to the old `<column>` with `TEXT` JSON
- Run `UPDATE <table> SET <column>_new = jsonb(<column>)`
- Drop the old column and rename the new column to the old name in one update

From there, I confirmed the many views and triggers I’ve defined (some of which iterate through the JSON) work without alteration. Indeed, all the previous SQLite JSON functions are documented to work seamlessly whether the JSON is `BLOB` or `TEXT`.

The only `JSONB`-specific command I used was [jsonb()](https://sqlite.org/json1.html#jminib) for the conversion. I see no need for it in regular data access.

## Code changes

Those manual SQLite changes were only for testing. The live database is actually built by GRDB, of course. I mention them just to paint a fuller picture of the difference.

In code, all I needed to do was change column types from `.text` to `.blob`. The types I persist as JSON already conform to `DatabaseValueConvertible` so that was it!

## Minor notes

I had SQL check constraints on the JSON columns that I had to  update from `json_valid(<column>) = 1` to `json_valid(<column>, 4) = 1`. The [second parameter](https://www.sqlite.org/json1.html#jvalid) indicates JSONB.

As I write this up, I’m a little worried I’ve overlooked some private extension in my code that made this easy. I don’t think so but if anything comes up, I’ll elaborate here!

**Comment by groue** [comment]
Thank you @Jason-Abbott for raising the topic of JSONB. I'll probably ask a few questions in the next few days :-)

**Comment by Jason-Abbott** [comment]
As an oldie who has been writing SQL for three decades 😄, let me plug [SQLiteFlow](https://www.sqliteflow.com) (no affiliation) as a fast and focused tool with JSONB support making it easy to still see the content of these BLOBs.

In fact, this is a nicer view of the JSON than before. Here I updated the `text` column to JSONB as described above.

<img width="663" alt="image" src="https://github.com/user-attachments/assets/29ef26a6-37ca-4d83-8c8f-dfb11da04892">

**Comment by Jason-Abbott** [comment]
I have worked a bit today on streamlining the JSONB encoding to avoid the need for string conversions and custom `DatabaseValueConvertible` conformance.

One hang-up happens here in `DatabaseValueEncoder`. That `value.encode()` always throws down to the `String` version for my basic `Codable` (but not explicitly `DatabaseValueConvertible`) types.

I *did* have some private methods that were glossing over these gaps. I think GRDB will need some signal to determine the `DatabaseValue.Storage` type for basic `Codable`s. 

Knowing the target column type would be a “magical” signal but may not always be feasible or performant. I’ll play around with some options here ...

```swift
func encode<T: Encodable>(_ value: T) throws {
  do {
  try value.encode(to: self)
      if requiresJSON {
          // Here we handle empty arrays and dictionaries.
          throw JSONRequiredError()
      }
    } catch is JSONRequiredError {
        let jsonData = try jsonEncoder.encode(value)
        
        // Store JSON String in the database for easier debugging and
        // database inspection. Thanks to SQLite weak typing, we won’t
        // have any trouble decoding this string into data when we
        // eventually perform JSON decoding.
        // TODO: possible optimization: avoid this conversion to string,
        // and store raw data bytes as an SQLite string
        let jsonString = String(decoding: jsonData, as: UTF8.self)
        try jsonString.encode(to: self)
    }
}
```

**Comment by Jason-Abbott** [comment]
**Edit: This isn’t actually valid. See the following comment.**

I made a quick solution that I’ll probably run with for a bit while I focus on other things. I updated the `Codable` types I want to store as JSONB to instead conform to `JSONBConvertible`, below. It handles everything, I think. This was quick and dirty so no criticism 🙂

```swift
public import GRDB
import OSLog

private let log = Logger.current

public protocol JSONBConvertible: Codable, DatabaseValueConvertible, Sendable {}

public extension JSONBConvertible {
    var databaseValue: DatabaseValue {
        let encoder = DatabaseValue.databaseJSONEncoder()
        do {
            let data = try encoder.encode(self)
            return data.databaseValue
        } catch {
            log.error("Failed to JSONEncode \(Self.Type.self)")
            return DatabaseValue.null
        }
    }

    static func fromDatabaseValue(_ dbValue: DatabaseValue) -> Self? {
        if case let .blob(data) = dbValue.storage {
            let decoder = DatabaseValue.databaseJSONDecoder()
            do {
                return try decoder.decode(self.self, from: data)
            } catch {
                log.error("Failed to JSONDecode \(Self.Type.self)")
                return nil
            }
        } else {
            return nil
        }
    }
}
```

**Comment by Jason-Abbott** [comment]
More learnings. It turns out most of the above was working thanks to a [bug in SQLite](https://sqlite.org/json1.html#the_json_blob_input_bug) that allows most things JSONish to function as `JSON` in a `BLOB` column even if it isn’t `JSONB`.

The test database conversion I did with `jsonb()` created valid `JSONB` but SQLite doesn’t do anything to automatically convert `JSON` bytes to `JSONB` upon insert.

The options to make that happen are (in a table where `value` is a `BLOB` intended to store `JSONB`) either
```sql
INSERT INTO jsonb_test (value) VALUES (jsonb('{"key1": 1, "key2": "two"}'))
INSERT INTO jsonb_test (value) VALUES (jsonb(x'7b226b657931223a20312c20226b657932223a202274776f227d'))
```

This is a little tricky to accomplish through GRDB helper functions since an attempt to insert the column value
```sql
jsonb('{"key1": 1, "key2": "two"}')
```
is emitted as (note extra quotes)
```sql
'jsonb(''{"key1": 1, "key2": “two"}'')'
```
which isn’t valid.

Obviously manual SQL will do the trick but I’ve been poking around for something more idiomatic.

This isn’t my top concern so commentary on the topic is likely to remain sporadic 🙂

**Comment by Jason-Abbott** [comment]
Last update for the day: Rather than try to automate something on the binding end of things, it works fine if the statement takes the form
```sql
INSERT INTO table (column) VALUES (jsonb(?))
```
But once made proper `JSONB`, it no longer decodes “for free.” Instead, the SELECT must assume the form
```sql
SELECT json(column) FROM table
```
which converts SQLite proprietary `JSONB` to `JSON` so `JSONDecoder` can process it.

At this point, the effort seems to exceed the benefit. In my own case, CloudKit is also hitting these tables so I would need to be careful to `jsonb()`/`json()` roundtrip every one of those changes. 😵‍💫

I am going to sleep on it but my inclination is to revert the changes I made, stick with `TEXT` `JSON`, and await a time, if ever, when SQLite offers some interfaces that automate the conversions.

I was enthusiastic for a little speed bump and size savings with `JSONB` but the steps to make it happen feel too brittle for my comfort.

**Comment by Jason-Abbott** [comment]
Last monologue update 🙂

The other way to make `JSONB` seamless, besides SQLite doing some kind of auto-conversion, is to implement `JSONBEncoder` and `JSONBDecoder` conforming to Swift `Encoder` and `Decoder`.

So I did that.

Looking at an existing [Rust version of the same](https://github.com/zamazan4ik/serde-sqlite-jsonb), and a [Swift binary encoder](https://github.com/Flight-School/MessagePack) for a different format, gave me a good start.

I am glad to share code if there’s interest (no signs of that yet 🙂) but since its currently tuned and tested for only my exact `Codable` types, I’m not going to put up a gist or anything yet.

Having looked at the Swift source for [JSONEncoder](https://github.com/swiftlang/swift-foundation/blob/main/Sources/FoundationEssentials/JSON/JSONEncoder.swift) and Decoder as part of this project, I think a direct binary encoder like this offers a lot of simplification.

**Comment by hekuli** [comment]
Hi, Just wondering what the status and/or best-practices for JSONB support is?

I'm building a new app and need to store large JSON BLOBs in my DB. I'd like to get the performance improvements of JSONB (but only if it's in a stable GRDB release).

The GRDB docs say: _"SQLite JSON functions and operators are documented to throw errors if any of their arguments are binary blobs. That’s the reason why it is recommended to store JSON as text."_

Yet when defining columns both options exist `.jsonText` and `.jsonb`. Just from reading the docs I'm still confused about which I should use and what the tradeoffs might be. Can anyone point me in the right direction? Thanks in advance!

**Reply by groue** [reply]
Hello @hekuli,

The GRDB doc is slightly misleading, for two reasons: JSONB is not covered (the doc assumes JSON strings in the database), and SQLite does not behaves as its own documentation says (JSON strings stored as data blobs do not throw errors).

> I'd like to get the performance improvements of JSONB (but only if it's in a stable GRDB release).

Not many people use JSONB, so it is not battle-tested. It is stable nonetheless. A [custom SQLite build](https://github.com/groue/GRDB.swift/blob/master/Documentation/CustomSQLiteBuilds.md) is required, because the SQLite version that ships with the operating system has no support for JSONB (...yet, unless I'm mistaken).

> Yet when defining columns both options exist `.jsonText` and `.jsonb`. Just from reading the docs I'm still confused about which I should use and what the tradeoffs might be. Can anyone point me in the right direction? Thanks in advance!

**When you define the schema**, prefer `jsonText` when you intend to store JSON text, and `jsonb` when you intend to store JSONB blobs. The choice here is more a matter of documenting your schema.

**When it is time to insert JSON text or JSON blobs in the database**, GRDB convenience apis are currently only able to help you store JSON text. To insert JSONB, you'll have to perform raw SQL, and provide a JSONB value.

To provide this JSONB values, your can build some JSON data, and call the `JSONB` SQL function:

```swift
let id: String = ...
let jsonData: Data = ...
try db.execute(literal: """
    INSERT INTO player (id, info)
    VALUES (\(id), JSONB(\(jsonData))
    """)
```

You can also have a look at https://github.com/Jason-Abbott/swift-sqlite-jsonb by @Jason-Abbott, which builds JSONB without the need to call the `JSONB` SQL function.

**When you want to convert JSONB database values into JSON data that your app can understand**, you need to perform the reverse conversion, with the `JSON` SQL function.

For example:

```swift
// SELECT id, JSON(info) FROM player
Player.select(Column("id"), Database.json(Column("info")))
```

Alternatively:

```swift
struct Player {
    var id: String

    // JSON data property
    var info: Data // JSON data

    // Also possible: a value that can be decoded from JSON
    var info: MyDecodedValue // Decodable
}

extension Player: Decodable, FetchableRecord, TableRecord {
    enum Columns {
        static let id = Column(CodingKeys.id)
        static let info = JSONColumn(CodingKeys.info)
    }
    static var databaseSelection: [any SQLSelectable] {
        [Columns.id, Database.json(Columns.info)]
    }
}

let players = try dbQueue.read { db in
    // SELECT id, JSON(info) FROM player
    try Player.fetchAll(db)
}
```

To avoid the intermediary JSON value in the above sample code, you'll again need a JSONB decoder such as the repo linked above, able to decode raw JSONB data. You'll also need to give up with the `Codable` convenience, and perform custom decoding in `FetchableRecord.init(row:)`.

**Support for JSONB is very early.** I'm only aware of one GRDB user using it (without known issue, but with elbow grease). Please spend some time exploring, trying to come up with your preferred setup, with patience. Share your experience when you're happy!

**Reply by hekuli** [reply]
Thanks so much @groue for the speedy and detailed reply!

JSONB sounds a bit too green and involved for my current comfort level. I'll just proceed with Text fields for now and revisit this once it's matured a bit.

Also I must say that I'm really loving working w/GRDB so far. Amazing project. Great work!

**Reply by groue** [reply]
Thank you, @hekuli! I suppose JSONB support will improve over time, so stay tuned :-)
